pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,364 @@
+	}
+	sfs := Service{
+		basic:             basic,
+		log:               logger,
+		cfg:               cfg,
+		chain:             chain,
+		shutdownCallback:  shutdownCallback,
+		stateSyncInterval: uint32(stateSyncInterval),
+	}
+	if sfs.stateSyncInterval <= 0 {
+		sfs.stateSyncInterval = config.DefaultStateSyncInterval
+	}
+	var (
+		containerID  cid.ID
+		containerObj container.Container
+	)
+	sfs.basic.Ctx, sfs.basic.CtxCancel = context.WithCancel(context.Background())
+	if err = sfs.basic.Pool.Dial(context.Background()); err != nil {
+		sfs.isShutdown.CompareAndSwap(false, true)
+		return nil, fmt.Errorf("failed to dial NeoFS pool: %w", err)
+	}
+
+	err = containerID.DecodeString(sfs.cfg.ContainerID)
+	if err != nil {
+		sfs.isShutdown.CompareAndSwap(false, true)
+		return nil, fmt.Errorf("failed to decode container ID: %w", err)
+	}
+
+	err = sfs.basic.Retry(func() error {
+		containerObj, err = sfs.basic.Pool.ContainerGet(sfs.basic.Ctx, containerID, client.PrmContainerGet{})
+		return err
+	})
+	if err != nil {
+		sfs.isShutdown.CompareAndSwap(false, true)
+		return nil, fmt.Errorf("failed to get container: %w", err)
+	}
+	containerMagic := containerObj.Attribute("Magic")
+	if containerMagic != strconv.Itoa(int(sfs.chain.GetConfig().Magic)) {
+		sfs.isShutdown.CompareAndSwap(false, true)
+		return nil, fmt.Errorf("container magic mismatch: expected %d, got %s", sfs.chain.GetConfig().Magic, containerMagic)
+	}
+	sfs.isShutdown.CompareAndSwap(true, false)
+	return &sfs, nil
+}
+
+// LastStateIndex returns the last state index from NeoFS.
+func (sfs *Service) LastStateIndex() (uint32, error) {
+	if sfs.lastStateIndex != 0 {
+		return sfs.lastStateIndex, nil
+	}
+	startIndex := uint32(1)
+	for {
+		prm := client.PrmObjectSearch{}
+		filters := object.NewSearchFilters()
+		filters.AddFilter(sfs.cfg.StateAttribute, fmt.Sprintf("%d", startIndex*sfs.stateSyncInterval), object.MatchStringEqual)
+		prm.SetFilters(filters)
+
+		ctx, cancel := context.WithTimeout(sfs.basic.Ctx, sfs.cfg.Timeout)
+		var (
+			blockOidsObject []oid.ID
+			err             error
+		)
+		err = sfs.basic.Retry(func() error {
+			blockOidsObject, err = neofs.ObjectSearch(ctx, sfs.basic.Pool, sfs.basic.Account.PrivateKey(), sfs.cfg.ContainerID, prm)
+			return err
+		})
+		cancel()
+		if err != nil {
+			sfs.isShutdown.CompareAndSwap(false, true)
+			return 0, fmt.Errorf("failed to find '%s' object with index %d: %w", sfs.cfg.StateAttribute, startIndex, err)
+		}
+		if len(blockOidsObject) == 0 {
+			sfs.lastStateIndex = (startIndex - 1) * sfs.stateSyncInterval
+			return sfs.lastStateIndex, nil
+		} else {
+			sfs.lastStateOid = blockOidsObject[0]
+		}
+		startIndex++
+	}
+}
+
+// Start begins state fetching.
+func (sfs *Service) Start() error {
+	if sfs.IsShutdown() {
+		return errors.New("service shut down")
+	}
+	if !sfs.isActive.CompareAndSwap(false, true) {
+		return nil
+	}
+	sfs.log.Info("state fetcher started")
+	sfs.wg.Add(1)
+	go sfs.run()
```

Comment:
You don't need a waitgroup for it since it's a single routine. Use a channel instead, we have a pattern of `runToExit` channels, take a look at how this pattern is implemented in other services.